src/backend/gpopt/translate/CQueryMutators.cpp

```
@@ -443,162 +443,6 @@ CQueryMutators::NeedsLevelsUpCorrection
-	{
-		return (Node *) gpdb::CopyObject(node);
-	}
-
-
-	if (IsA(node, Var) && context->m_is_mutating_agg_arg)
-	{
-		// if we are mutating an aggregate argument do nothing since the aggregate will be place in the derived table's target list
-		// fix any outer references in the grouping column expression or arguments of an aggregate
-		return (Node *) IncrLevelsUpInVar((Var*) node);
-	}
-	
-	// if we find an aggregate or precentile expression then insert into the new derived table
-	// and refer to it in the top-level query
-
-	if (IsA(node, Aggref))
-	{
-		Aggref *old_aggref = (Aggref*) node;
-		Aggref *aggref = FlatCopyAggref(old_aggref);
-		aggref->agglevelsup = old_aggref->agglevelsup;
-
-		List *new_args = NIL;
-		ListCell *lc = NULL;
-
-		BOOL fAggregate = context->m_is_mutating_agg_arg;
-		context->m_is_mutating_agg_arg = true;
-
-		ForEach (lc, old_aggref->args)
-		{
-			Node *arg = (Node *) gpdb::CopyObject((Node*) lfirst(lc));
-			GPOS_ASSERT(NULL != arg);
-			// traverse each argument and fix levels up when needed
-			new_args = gpdb::LAppend
-						(
-						new_args,
-						gpdb::MutateQueryOrExpressionTree
-							(
-							arg,
-							(MutatorWalkerFn) CQueryMutators::RunGroupByProjListMutator,
-							(void *) context,
-							0 // flags -- mutate into cte-lists
-							)
-						);
-		}
-		context->m_is_mutating_agg_arg = fAggregate;
-		aggref->args = new_args;
-
-		const ULONG attno = gpdb::ListLength(context->m_groupby_target_list) + 1;
-		TargetEntry *target_entry = GetTargetEntryForAggExpr(context->m_mp, context->m_md_accessor, (Node *) aggref, attno);
-
-		// Add a new target entry to the query
-		context->m_groupby_target_list = gpdb::LAppend(context->m_groupby_target_list, target_entry);
-
-		Var *new_var = gpdb::MakeVar
-						(
-						1, // varno
-						(AttrNumber) attno,
-						gpdb::ExprType(node),
-						gpdb::ExprTypeMod(node),
-						0 // query levelsup
-						);
-
-		return (Node*) new_var;
-	}
-
-	if (IsA(node, GroupingFunc))
```

is handling for GroupingFunc not required? The method `RunExtractAggregatesMutator` doesn't handle it.

---

Note: For 5X, This func handles PercentileExpr as well.

---

> is handling for GroupingFunc not required (for levels other than 0)? The method RunExtractAggregatesMutator doesn't handle it.

Yeah this applies for PercentileFunc in 5x. Short answer, I don't think it is required, but I'm not 100% sure. But, it wasn't supported before. So in that sense there has not been any change. Let me think about this more, and maybe write a test case for it.

> Note: For 5X, This func handles PercentileExpr as well. so the code for 5x will be slightly different?

Yup - master/6x got rid of PercentileExpr

---

Ok updates on the GroupingFunc. We need to handle it even if it is nested. I think I have a potential test case.

---

Here is the repro of the issue:
```
create table test (a int, b int);
insert into test values (1, 1), (null, 1);

create table jazz (g int, h int, i int);
insert into jazz values (2, 2, 2);


shardikar=#  select a,sum(b), (select grouping(a) from jazz group by test.a, jazz.g) from test group by rollup(a);
 a | sum | grouping
---+-----+----------
 1 |   1 |        0
   |   1 |        0
   |   2 |        0
(3 rows)
```

Postgres results:
```
postgres=# select a,sum(b), (select grouping(a) from jazz group by test.a, jazz.g) from test group by rollup(a);
 a | sum | grouping
---+-----+----------
   |   2 |        1
   |   1 |        0
 1 |   1 |        0
(3 rows)
```

But, unfortunately, we can't do anything about it! The reason is that `struct GroupingFunc` in GPDB doesn't track the levels up information. Compare that to upstream [postgres](https://github.com/postgres/postgres/blob/3b925e905de3204ffef64fa4d53dd7bbac1a143f/src/include/nodes/primnodes.h#L342). 

So, I guess we must wait till the relevant [commit](https://github.com/postgres/postgres/commit/f3d3118532175541a9a96ed78881a3b04a057128) is merge into GPDB. Now, there are ways around this, say by iterating `GroupingFunc::args`, but I don't think that would belong in the ORCA normalization.

---

So, should we fallback in such cases as it gives wrong results?

---

If we fall back, the best case we get a wrong result (5x) or crash (6x). See https://github.com/greenplum-db/gpdb/issues/7279. Personally I don't think we should do anything till GPDB can support that kind of query.

Well, maybe we can throw an error during parsing. Let me mention that on the PR.